publication/2019-10-21_announcing_ngxs_3_6/article.md

```diff
@@ -0,0 +1,595 @@
+
+#### Here are a couple of example use cases:
+
+I. A convenient way to track state changes:
+
+_Before_
+
+```ts
+@State({ .. })
+class MyState {}
+
+@Component({ })
+class MyComponent {
+  constructor(store: Store) {
+    store.select(MyState).subscribe((newState) => {
+       console.log('state is changed', newState);
+    })
+  }
+}
+```
+
+One of the problems is that if we are not using the `@ngxs/logger-plugin` or `@ngxs/devtools-plugin`, then we do not know what the previous state was before our state changed. It's great to have such an opportunity out of the box for quick, simple and focused debugging.
+
+_After_
+
+```ts
+@State({ .. })
+class MyState implements NgxsOnChanges {
+  public ngxsOnChanges({ previousValue, currentValue }: NgxsSimpleChange): void {
+    console.log('MyState has changed', previousValue, currentValue);
+  }
+}
+```
+
+Nice!
+
+II. Convenient to synchronize with the server
+
+Sometimes we need to save state on the server every time the client makes changes to it.
+
+_Before_
+
+```ts
+@State({ .. })
+class MyState {}
+
+@Component({})
+class MyComponent {
+  constructor(store: Store, api: ApiService) {
+    store.select(MyState).subscribe(async newState => {
+      await api.save(newState);
+    });
+  }
+}
+```
+
+_After_
+
+```ts
+@State({ .. })
+class MyState implements NgxsOnChanges {
+  constructor(api: ApiService) {}
+  public async ngxsOnChanges(change: NgxsSimpleChange): void {
+    await api.save(change.currentValue);
+  }
+}
+```
+
+III. You can write your own custom logger without another plugins
+
+```ts
+@State({ .. })
+class MyState implements NgxsOnChanges {
+  public ngxsOnChanges({ previousValue, currentValue }: NgxsSimpleChange): void {
+    console.log('prev state', previousValue);
+    console.log('next state', currentValue);
+  }
+}
+```
+
+## 💦 Fixed Actions Stream Subscriptions Leak
+
+[#1381](https://github.com/ngxs/store/pull/1381)
+(Introduction [with problem statement], details and usage)
+
+## 🚧 Improved type safety for children states
+
+[#1388](https://github.com/ngxs/store/pull/1388)
+(Introduction [with problem statement], details and usage)
+
+## 🐛 Bug Fixes
+
+For Each:
+(Introduction, details and usage)
+
+- Fix: Explicit typings for state operators [#1395](https://github.com/ngxs/store/pull/1395)
+- Fix: Warn if the zone is not actual "NgZone" [#1270](https://github.com/ngxs/store/pull/1270)
+- Fix: Do not re-throw error to the global handler if custom is provided [#1379](https://github.com/ngxs/store/pull/1379)
+
+## 🔌 Plugin Improvements and Fixes
+
+### HMR Plugin
+
+- Feature: HMR Plugin - Add option for persisting state after the root module is disposed [#1369](https://github.com/ngxs/store/pull/1369)
+
+If you make any changes to state in your application during the `ngOnDestroy` Angular lifecycle hook on any of your primary application components then it those changes may have missed the opportunity to be saved for Hot Module Reloading:
```

Comment:
@splincode I adjusted the description here because I was battling to understand what you were meaning by side effects. Is my adapted description correct for this change (read the whole section)?

Reply:
No, the new flag just ignores changes during `ngOnDestroy`. Because `ngOnDestroy` could incorrectly change the state, which simply should have been used in the next hmr cycle.

Example HMR cycle
1. Initial first app (no hmr changes)

```
appState {
   A: {
      B: {
          value: []
       }
    },
    C: {
       value: []
    }
}
```

2. Mutate app state in app (dispatched any actions)

new Action1(),
new Action2(),
...
new ActionN()

We have

```
appState {
   A: {
      B: {
          value: [1, 2, 3]
       }
    },
    C: {
       value: [4, 5, 6]
    }
}
```

3. Changes typescript file for emit hot reload

Moreover, if we have such a code

```ts
class BComponent {
   ngOnDestroy() {
      // reset B state (set value [])
   }
}
```

```ts
class CComponent {
  ngOnDestroy() {
      // reset C state (set value [])
   }
}
```

After tiny refreshing the page with HMR Cycle

  a) Without persistAfterDestroy 

We have

```
appState {
   A: {
      B: {
          value: []
       }
    },
    C: {
       value: []
    }
}
```

  b) With persistAfterDestroy 


```
appState {
   A: {
      B: {
          value: [1, 2, 3]
       }
    },
    C: {
       value: [4, 5, 6]
    }
}
```

Reply:
This is a good description.

```
this.store.reset({ .. }); // These changes will not be remembered by HMR
```

Reply:
Hmmm... I am wondering if `persistAfterDestroy` is a good name for the setting then?
I was not able to infer that behaviour from the name.
What is meant by `persist`? I interpreted it as **store** the state after destroy.
Let's try to think of some better names...

Reply:
Maybe:
`ignoreNgDestroy` - ignores state changes made in `ngDestroy`
`persistBeforeDestroy` - state is remembered for the hmr reload before the angular app runs its' destory lifecycle hooks.
Any other ideas?
I think being able to list a name that makes sense with a single sentence description is a good test.